### PR TITLE
Fixation!

### PR DESCRIPTION
**Pull request by adamemerson**
A couple minor fixes, one for compilation, one for compatibility.

### CODE REVIEW

File: src/rgw/rgw_iam_policy.cc [hunk 1]
```diff
@@ -3,11 +3,12 @@
 
 
 #include <cstring>
-#include <regex>
 #include <sstream>
 #include <stack>
 #include <utility>
 
+#include <boost/regex.hpp>
```
Comment (tchaikov) [inline src/rgw/rgw_iam_policy.cc]: this commit addresses http://tracker.ceph.com/issues/20112.

File: src/rgw/rgw_iam_policy.cc [hunk 2]
```diff
@@ -202,12 +205,9 @@ ARN::ARN(const rgw_bucket& b, const string& o)
 
 optional<ARN> ARN::parse(const string& s, bool wildcards) {
   static const regex rx_wild("arn:([^:]*):([^:]*):([^:]*):([^:]*):([^:]*)",
-			     std::regex_constants::ECMAScript |
-			     std::regex_constants::optimize);
+			     ECMAScript | optimize);
```
Comment (tchaikov) [inline src/rgw/rgw_iam_policy.cc]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_iam_policy.cc:210:78: error: call of overloaded ‘basic_regex(const char [49], int)’ is ambiguous
     "arn:([^:*]*):([^:*]*):([^:*]*):([^:*]*):([^:*]*)", ECMAScript | optimize);
                                                                              ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/regex/v4/regex.hpp:67:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/regex.hpp:31,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_iam_policy.cc:10:
/home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/regex/v4/basic_regex.hpp:337:4: note: candidate: boost::basic_regex<charT, traits>::basic_regex(const charT*, const charT*, boost::basic_regex<charT, traits>::flag_type) [with charT = char; traits = boost::regex_traits<char>; boost::basic_regex<charT, traits>::flag_type = unsigned int]
    basic_regex(const charT* p1, const charT* p2, flag_type f = regex_constants::normal)
    ^
/home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/regex/v4/basic_regex.hpp:333:13: note: candidate: boost::basic_regex<charT, traits>::basic_regex(const charT*, boost::basic_regex<charT, traits>::flag_type) [with charT = char; traits = boost::regex_traits<char>; boost::basic_regex<charT, traits>::flag_type = unsigned int]
    explicit basic_regex(const charT* p, flag_type f = regex_constants::normal)
             ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_iam_policy.cc: In function ‘boost::optional<rgw::auth::Principal> rgw::IAM::parse_principal(CephContext*, rgw::IAM::TokenID, std::__cxx11::string&&)’:
```

i guess, the type of OR'ed enums can be taken as `unsigned int` or `const char*` by GCC 4:5.3.1-1ubuntu1.  we might need to cast it to `regex::flag_type` explicitly.

**Comment by adamemerson** [comment]
What distribution uses that? Because if it's Trusty Tahr, we should
kick it into a ditch and bury it and have a party and dance on its
grave. Canonical doesn't support it; I don't think we should.

**Comment by adamemerson** [comment]
Also I think this might actually be caused by the horrible using
namespace std we have everywhere. Let me try being explicit just to
see if that fixes it.

**Comment by tchaikov** [comment]
@adamemerson

i think that's xenial: https://packages.ubuntu.com/xenial/gcc

File: src/mon/FSCommands.cc [hunk 3]
```diff
@@ -302,12 +302,12 @@ class SetHandler : public FileSystemCommandHandler
         ss << "unsetting the metadata load balancer";
       else 
         ss << "setting the metadata load balancer to " << val;
-        fsmap.modify_filesystem(
-            fs->fscid,
-            [val](std::shared_ptr<Filesystem> fs)
-        {
-          fs->mds_map.set_balancer(val);
-        });
+      fsmap.modify_filesystem(
```
Comment (tchaikov) [inline src/mon/FSCommands.cc]: this has been fixed in master.

File: src/common/config_opts.h [hunk 4]
```diff
@@ -1135,7 +1135,7 @@ OPTION(bluestore_allocator, OPT_STR, "bitmap")     // stupid | bitmap
 OPTION(bluestore_freelist_blocks_per_key, OPT_INT, 128)
 OPTION(bluestore_bitmapallocator_blocks_per_zone, OPT_INT, 1024) // must be power of 2 aligned, e.g., 512, 1024, 2048...
 OPTION(bluestore_bitmapallocator_span_size, OPT_INT, 1024) // must be power of 2 aligned, e.g., 512, 1024, 2048...
-OPTION(bluestore_max_deferred_txc, OPT_INT, 32)
+OPTION(bluestore_max_deferred_txc, OPT_U32, 32)
```
Comment (tchaikov) [inline src/common/config_opts.h]: this has been fixed in master.

File: src/rgw/rgw_iam_policy.cc [hunk 5]
```diff
@@ -201,19 +199,23 @@ ARN::ARN(const rgw_bucket& b, const string& o)
...
+				       sizeof(str_no_wild) - 1,
+				       boost::regex_constants::ECMAScript |
+				       boost::regex_constants::optimize);
+
+  boost::smatch match;
 
   if ((s == "*") && wildcards) {
     return ARN(Partition::wildcard, Service::wildcard, "*", "*", "*");
-  } else if (regex_match(s, match, wildcards ? rx_wild : rx_no_wild)) {
+  } else if (boost::regex_match(s, match, wildcards ? rx_wild : rx_no_wild)) {
```
Comment (tchaikov) [inline src/rgw/rgw_iam_policy.cc]: BTW, i think it's fine to use `using boost::regex_match` here

File: src/rgw/rgw_iam_policy.cc (hunk 1 above)
Comment (liupan1111) [inline src/rgw/rgw_iam_policy.cc]: cool, thanks for remind.

**Pull request merged**